statsmodels/base/_penalized.py

```diff
@@ -0,0 +1,194 @@
+        If None, then pen_weight is set to nobs.
+
+
+    TODO: missing **kwds or explicit keywords
+
+    TODO: do we adjust the inherited docstrings?
+    We would need templating to add the penalization parameters
+
+    """
+
+    def __init__(self, *args, **kwds):
+
+        # pop extra kwds before calling super
+        self.penal = kwds.pop('penal', None)
+        self.pen_weight =  kwds.pop('pen_weight', None)
+
+        super(PenalizedMixin, self).__init__(*args, **kwds)
+
+        # TODO: define pen_weight as average pen_weight? i.e. per observation
+        # I would have prefered len(self.endog) * kwds.get('pen_weight', 1)
+        # or use pen_weight_factor in signature
+        if self.pen_weight is None:
+            self.pen_weight = len(self.endog)
+
+        if self.penal is None:
+            # unpenalized by default
+            self.penal = NonePenalty()
+            self.pen_weight = 0
+
+        self._init_keys.extend(['penal', 'pen_weight'])
+
+    def loglike(self, params, pen_weight=None, **kwds):
+        if pen_weight is None:
+            pen_weight = self.pen_weight
+
+        llf = super(PenalizedMixin, self).loglike(params, **kwds)
+        if pen_weight != 0:
+            llf -= pen_weight * self.penal.func(params)
+
+        return llf
+
+
+    def loglikeobs(self, params, pen_weight=None, **kwds):
+        if pen_weight is None:
+            pen_weight = self.pen_weight
+
+        llf = super(PenalizedMixin, self).loglikeobs(params, **kwds)
+        nobs_llf = float(llf.shape[0])
+
+        if pen_weight != 0:
+            llf -= pen_weight / nobs_llf * self.penal.func(params)
+
+        return llf
+
+    def score_numdiff(self, params, pen_weight=None, method='fd', **kwds):
+        """score based on finite difference derivative
+
+        """
+        if pen_weight is None:
+            pen_weight = self.pen_weight
+
+        loglike = lambda p: self.loglike(p, pen_weight=pen_weight, **kwds)
+
+        if method == 'cs':
+            return approx_fprime_cs(params, loglike)
+        elif method == 'fd':
+            return approx_fprime(params, loglike, centered=True)
+        else:
+            raise ValueError('method not recognize, should be "fd" or "cs"')
+
+    def score(self, params, pen_weight=None, **kwds):
+        if pen_weight is None:
+            pen_weight = self.pen_weight
+
+        sc = super(PenalizedMixin, self).score(params, **kwds)
+        if pen_weight != 0:
+            sc -= pen_weight * self.penal.grad(params)
+
+        return sc
+
+    def score_obs(self, params, pen_weight=None, **kwargs):
+        if pen_weight is None:
+            pen_weight = self.pen_weight
+
+        sc = super(PenalizedMixin, self).score_obs(params, **kwargs)
+        nobs_sc = float(sc.shape[0])
+        if pen_weight != 0:
+            sc -= pen_weight / nobs_sc  * self.penal.grad(params)
+
+        return sc
+
+    def hessian_numdiff(self, params, pen_weight=None, **kwds):
```

Comment:
Does this need to be part of the public API, or is it analogous to e.g. `NegativeBinomial._score_nb1`?

Reply:
I keep it public. It is convenient to check new penalty functions that we or users can add.
(NegativeBinomial is different because it dispatches score to one of the `_score_xxx`)